File: f90/src/findFractionOfPixelWithinCircle.f90
PURE SUBROUTINE findFractionOfPixelWithinCircle(ndiv, xmin, xmax, ymin, ymax, r, cx, cy, frac)
    ! Import standard modules ...
    USE ISO_C_BINDING

    IMPLICIT NONE

    ! Declare inputs/outputs ...
    INTEGER(kind = C_LONG_LONG), INTENT(in)                                     :: ndiv
    REAL(kind = C_DOUBLE), INTENT(in)                                           :: xmin
    REAL(kind = C_DOUBLE), INTENT(in)                                           :: xmax
    REAL(kind = C_DOUBLE), INTENT(in)                                           :: ymin
    REAL(kind = C_DOUBLE), INTENT(in)                                           :: ymax
    REAL(kind = C_DOUBLE), INTENT(in)                                           :: r
    REAL(kind = C_DOUBLE), INTENT(in)                                           :: cx
    REAL(kind = C_DOUBLE), INTENT(in)                                           :: cy
    REAL(kind = C_DOUBLE), INTENT(out)                                          :: frac

    ! Declare internal variables ...
    INTEGER(kind = C_LONG_LONG)                                                 :: ix
    INTEGER(kind = C_LONG_LONG)                                                 :: iy
    REAL(kind = C_DOUBLE)                                                       :: dx
    REAL(kind = C_DOUBLE)                                                       :: dy
    REAL(kind = C_DOUBLE), DIMENSION(ndiv)                                      :: xaxis
    REAL(kind = C_DOUBLE), DIMENSION(ndiv)                                      :: yaxis
    REAL(kind = C_DOUBLE), DIMENSION(ndiv, ndiv)                                :: dist

    ! Calculate size of pixels ...
    dx = (xmax - xmin) / REAL(ndiv, kind = C_DOUBLE)
    dy = (ymax - ymin) / REAL(ndiv, kind = C_DOUBLE)

    ! Create centroids relative to the centre of the circle ...
    DO ix = 1_C_LONG_LONG, ndiv
        xaxis(ix) = xmin + (REAL(ix, kind = C_DOUBLE) - 0.5e0_C_DOUBLE) * dx - cx
    END DO

    ! Create centroids relative to the centre of the circle ...
    DO iy = 1_C_LONG_LONG, ndiv
        yaxis(iy) = ymin + (REAL(iy, kind = C_DOUBLE) - 0.5e0_C_DOUBLE) * dy - cy
    END DO

    ! Find out the distance of each centroid to the centre of the circle ...
    DO ix = 1_C_LONG_LONG, ndiv
        DO iy = 1_C_LONG_LONG, ndiv
            dist(iy, ix) = HYPOT(xaxis(ix), yaxis(iy))
        END DO
    END DO

    ! Calculate answer ...
    frac = REAL(COUNT(dist <= r, kind = C_LONG_LONG), kind = C_DOUBLE) / REAL(ndiv * ndiv, kind = C_DOUBLE)
END SUBROUTINE findFractionOfPixelWithinCircle

File: hml/findFractionOfPixelWithinCircle.py
#!/usr/bin/env python3

# Define function ...
def findFractionOfPixelWithinCircle(xmin, xmax, ymin, ymax, r, kwArgCheck = None, cx = 0.0, cy = 0.0, ndiv = 16):
    """
    Find the fraction of a pixel that is within a hard circular mask.

    Arguments:
    xmin -- left edge
    xmax -- right edge
    ymin -- lower edge
    ymax -- upper edge
    r -- radius of circle

    Keyword arguments:
    cx -- x position of centre of circle (default 0.0)
    cy -- y position of centre of circle (default 0.0)
    ndiv -- number sub-divisions (default 16)

    Note:
    This function is crying out for FORTRAN+OpenMP.
    """

    # Import special modules ...
    try:
        import numpy
    except:
        raise Exception("\"numpy\" is not installed; run \"pip install --user numpy\"") from None

    # Check keyword arguments ...
    if kwArgCheck is not None:
        print(f"WARNING: \"{__name__}\" has been called with an extra positional argument")

    # Create nodes relative to the centre of the circle ...
    xaxis = numpy.linspace(xmin, xmax, num = ndiv + 1) - cx
    yaxis = numpy.linspace(ymin, ymax, num = ndiv + 1) - cy

    # Convert the nodes to centroids ...
    # NOTE: https://stackoverflow.com/a/23856065
    xaxis = 0.5 * (xaxis[1:] + xaxis[:-1])
    yaxis = 0.5 * (yaxis[1:] + yaxis[:-1])

    # Find out the distance of each centroid to the centre of the circle ...
    dist = numpy.zeros((ndiv, ndiv), dtype = numpy.float64)
    for ix in range(ndiv):
        for iy in range(ndiv):
            dist[iy, ix] = numpy.hypot(xaxis[ix], yaxis[iy])

    # Return answer ...
    return float((dist <= r).sum()) / float(ndiv * ndiv)
